Feat: Adds kongctl to developer (#4047)
Kong/developer.konghq.com: app/kongctl/declarative.md

---
title: Declarative configuration with kongctl

description: Learn how to manage {{site.konnect_product_name}} infrastructure as code using declarative YAML configuration.

content_type: reference
layout: reference

beta: true

works_on:
  - konnect

tools:
  - kongctl

breadcrumbs:
  - /kongctl/

related_resources:
  - text: CI/CD integration guide
    url: /kongctl/ci-cd/
  - text: Get started with kongctl
    url: /kongctl/get-started/
  - text: Supported resources
    url: /kongctl/supported-resources/
---

kongctl's declarative management feature enables you to manage {{site.konnect_short_name}} resources using simple YAML configuration files and a stateless CLI tool. 
This approach provides version control, automation, and predictable deployments through a plan-based workflow.

{:.info}
> **New to kongctl?** This document provides an extensive reference to declarative management with kongctl. For a shorter getting started experience, see 
the [Get started with kongctl](/kongctl/get-started/) guide to quickly learn basic usage and resource management.

> ## Overview

Reviewer: ```suggestion
{:.success}
> **New to kongctl?** This document provides an extensive reference to declarative management with kongctl. For a shorter getting started experience, see 
the [Get started with kongctl](/kongctl/get-started/) guide to quickly learn basic usage and resource management.

kongctl's declarative management feature enables you to manage {{site.konnect_short_name}} resources using simple YAML configuration files and a stateless CLI tool. 
This approach provides version control, automation, and predictable deployments through a plan-based workflow.

```

- **Plan-based workflow**: Generate plan artifacts that represent the changes needed, review them, and apply them later
- **Stateless operation**: kongctl queries live {{site.konnect_short_name}} state instead of maintaining a separate state file
- **Namespace ownership**: Use namespaces to group resources between teams and environments

## Core concepts

### Configuration input files

kongctl declarative configuration uses YAML files to describe your desired state. These files define resources and their properties using a simple, structured format.

#### Basic structure

```yaml
portals: 
  - ref: developer-portal
    name: "developer-portal"
    display_name: "Developer Portal"
  - ref: portal-num-two
    name: "another portal"
    display_name: "Amazing Dev Portal"

apis:
  - ref: users-api
    name: "Users API"
```

#### Multiple files

Different resources and resource types can be defined in a single file, or split across multiple files:

```
project/
├── portals.yaml
├── apis.yaml
└── auth-strategies.yaml
```

### Resource identity

> Resources in {{site.konnect_short_name}} and kongctl will have multiple identifiers:

Reviewer: ```suggestion
Resources in {{site.konnect_short_name}} and kongctl have multiple identifiers:
```

in order to manage the resource going forward.

## kongctl metadata

Within the input configuration, users can provide some kongctl specific metadata values that affects the behavior of the 
declarative system. This `kongctl` section can be provided on individual resources or at the configuration file level. 

These values are stored on resources via the {{site.konnect_short_name}} [resource labels](/konnect-platform/konnect-labels/),
so only resources supporting labels are supported. In general only parent resources support labels and child 
resources inherit the values from their parents.

### Namespace 

The `namespace` field allows the user to define ownership and isolation of resources. You may choose to assign any
`namespace` you choose, but common patterns would be teams (finance, engineering), environments (prod, dev), etc...

```yaml
apis:
  - ref: billing-api
    name: "Billing API"
    kongctl:
      namespace: finance-team
```

### Protected resources

When a resource is marked `protected: true`, the kongctl declarative planner will not allow planning of
updates or deletes to those resources. In order to update or delete these resources, only changing the

> resource from `protected: true` to `protected: false` is allowed before subsequent changes can be made.

Reviewer: ```suggestion
When a resource is marked `protected: true`, the kongctl declarative planner will not allow planning of
updates or deletes to those resources. To update or delete these resources, change the
resource from `protected: true` to `protected: false`. Once you update this, then you are allowed allowed to make subsequent changes.
```

```yaml
_defaults:
  kongctl:
    namespace: platform-team
    protected: true

portals:
  - ref: api-portal
    name: "API Portal"
    # Inherits: namespace=platform-team, protected=true

  - ref: test-portal
    name: "Test Portal"
    kongctl:
      # Overrides both defaults
      namespace: qa-team
      protected: false
```

### kongctl metadata precedence

**Namespace behavior**:

| File Default | Resource Value | Result | Notes |
|--------------|----------------|--------|-------|
| Not set | Not set | `default` | Default |
| Not set | `team-a` | `team-a` | Resource explicit |
| `team-b` | Not set | `team-b` | Inherits default |
| `team-b` | `team-a` | `team-a` | Resource overrides |

**Protected field behavior**:

| File Default | Resource Value | Result | Notes |
|--------------|----------------|--------|-------|
| Not set | Not set | `false` | Default |
| Not set | `true` | `true` | Resource explicit |
| `true` | Not set | `true` | Inherits default |
| `true` | `false` | `false` | Resource overrides |

**Child resources** automatically inherit the namespace from their parent.

### namespace enforcement flags

> The following flags allow further control over namespaces when running the declarative commands.

Reviewer: ```suggestion
The following flags allow further control over namespaces when running the declarative commands:
```

1. You provide desired state (YAML files)
2. kongctl queries current state from {{site.konnect_short_name}}
3. kongctl calculates the difference
4. kongctl applies changes

This means:
* No state file to manage or lock
* Always reflects live {{site.konnect_short_name}} state
* Can run from anywhere with the same configuration
* Multiple people can use the same configuration files

{:.warning}
> **Caution**: Be careful with concurrent operations on the same resources. Use namespace isolation to avoid conflicts.

## YAML tags

YAML tags act as pre-processors, allowing you to load content from external files and cross-reference resources.

### Loading file content

Use `!file` to load entire file contents. This is very useful with resources that have large content often stored in files
like OpenAPI specifications or Developer Portal pages and documentation.

```yaml
apis:
  - ref: users-api
    name: "Users API"
    description: !file ./docs/api-description.md
    versions:
      - ref: v1
        spec: !file ./specs/users-v1.yaml
```

### Extracting values from structured files

When the input file for `!file` is a structured file (`JSON` or `YAML`), you can

> EXtract specific values using `!file <path>#field` notation:

Reviewer: ```suggestion
extract specific values using the `!file <path>#field` notation:
```

```yaml
apis:
  - ref: users-api
    name: !file ./specs/openapi.yaml#info.title
    description: !file ./specs/openapi.yaml#info.description
    versions:
      - ref: v1
        spec: !file ./specs/openapi.yaml
```

### Referencing other resources

> Use the `!ref` yaml tag to reference values from other declared resources:

Reviewer: ```suggestion
Use the `!ref` YAML tag to reference values from other declared resources:
```

```yaml
control_planes:
  - ref: prod-us-runtime
    name: "prod-us"
    cluster_type: "CLUSTER_TYPE_K8S_INGRESS_CONTROLLER"

  - ref: shared-group
    name: "shared-group"
    cluster_type: "CLUSTER_TYPE_CONTROL_PLANE_GROUP"
    members:
      - id: !ref prod-us-runtime#id
```

### Path resolution and security

> **Path resolution**: All file paths are relative to the directory containing the configuration file.

Reviewer: ```suggestion
Path resolution is when all file paths are relative to the directory containing the configuration file.
```

> **Security features**:

Reviewer: ```suggestion
Path resolution has the following security features:
```

* Files limited to 10MB
* Default base directory is the parent of each `-f` source file
* Override with `--base-dir` flag

```yaml
# ✅ Allowed (within base directory)
description: !file ./docs/description.txt
spec: !file ../shared/openapi.yaml

# ❌ Blocked (absolute path)
config: !file /etc/passwd

# ❌ Blocked (outside base directory)
secret: !file ../../../sensitive.yaml
```

**File caching**: Files are cached during execution for performance:

```yaml
apis:
  - ref: api-1
    name: !file ./common.yaml#api.name        # Loaded and cached
    description: !file ./common.yaml#api.desc # Uses cache
  - ref: api-2
    team: !file ./common.yaml#team.name       # Uses cache
```

## External resources

The external resources feature allows you to reference {{site.konnect_short_name}} resources
managed outside of kongctl. Some {{site.konnect_short_name}} resources reference others, so this
can be useful for cross-team references, or references to resources created by different management techniques or 
technologies.

### Basic syntax

The resource requires a `ref` field, which is how it is further referenced by dependent resources. 

> The resource is marked external by marking it with an `_external` key. Under the `_external` key

Reviewer: ```suggestion
The resource is denoted as external by marking it with an `_external` key. Below the `_external` key,
```

you define a _selector_ which will query the organization for resources matching given fields.

```yaml
portals:
  - ref: shared-developer-portal
    _external:
      selector:
        matchFields:
          name: "Shared Developer Portal"
```

### Key characteristics

* **Cannot declare kongctl metadata**: External resources don't support `kongctl.namespace` or `kongctl.protected`
* **Not included in sync planning**: External namespaces don't affect deletion calculations
* **Used for references**: Child resources can reference external parents

### API Publication and shared portal example

```yaml
# External portal managed by platform team
portals:
  - ref: platform-portal
    _external:
      selector:
        matchLabels:
          team: platform

# Your API published to the external portal
api_publications:
  - ref: my-api-pub
    api: my-api
    portal_id: !ref platform-portal#id 
```

## Best practices

### Use namespaces for isolation

Separate resources by team or environment:

```yaml
_defaults:
  kongctl:
    namespace: team-alpha

apis:
  - ref: team-api
    name: "Team API"
    # Automatically in team-alpha namespace
```

> ### Protect critical resources

Reviewer: ```suggestion
Protect critical resources:
```

```yaml
apis:
  - ref: production-api
    name: "Production API"
    kongctl:
      namespace: production
      protected: true  # Prevents accidental deletion
```

### Review plans before applying

> Use the two-phase workflow for production:

Reviewer: ```suggestion
Review plans before applying. Use the two-phase workflow for production:
```

```bash
# Generate plan
kongctl plan -f config.yaml --output-file plan.json

# Review in pull request or approval process
kongctl diff --plan plan.json

# Apply after approval
kongctl apply --plan plan.json
```

## Examples

For complete working examples, see the [kongctl examples directory](https://github.com/Kong/kongctl/tree/main/docs/examples/declarative).

## Troubleshooting

### Authentication failures

```bash
# Verify authentication
kongctl get me

# Re-authenticate if needed
kongctl login
```

### Plan generation failures

* Validate YAML syntax
* Check file paths are correct and within base directory
* Verify network connectivity to {{site.konnect_short_name}}

### Apply failures

* Review the plan for conflicts
* Check for protected resources blocking deletion
* Verify all referenced resources exist

### File loading errors

```
Error: failed to process file tag: file not found

> ```

Reviewer: ```suggestion
If you are loading a file and get the `Error: failed to process file tag: file not found`, do the following:
```

* Verify file path is correct and relative to config file
* Check file exists and is within base directory
* Ensure file size is under 10MB

### Field validation errors

kongctl uses strict validation:

```yaml
portals:
  - ref: my-portal
    lables:  # ❌ ERROR: Unknown field. Did you mean 'labels'?
      team: platform
```

Enable debug logging:

```bash
kongctl apply -f config.yaml --log-level debug
```

## Next steps

* [See example declarative configurations](https://github.com/Kong/kongctl/tree/main/docs/examples/declarative)
* [Learn kongctl authorization options](/kongctl/authentication/)
* [kongctl configuration reference guide](/kongctl/config/) 
* [kongctl troubleshooting guide](/kongctl/troubleshooting/)
* [Using kongctl and deck for full API platform management](/kongctl/kongctl-and-deck/)

> * [View the {{site.konnect_short_name}} API reference](/konnect-api/)

Reviewer: ```suggestion
```
moving to the next steps frontmatter metadata
